diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 3dd88fcef..589146044 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5634,7 +5634,7 @@ static void hostapd_disassoc_sta(struct hostapd_data *hapd,
 
 
 #ifdef CONFIG_IEEE80211BE
-static struct sta_info *
+struct sta_info *
 hostapd_ml_get_assoc_sta(struct hostapd_data *hapd, struct sta_info *sta,
 			 struct hostapd_data **assoc_hapd)
 {
diff --git a/src/ap/ieee802_11.h b/src/ap/ieee802_11.h
index 4e7d8fbe8..e7b4639bd 100644
--- a/src/ap/ieee802_11.h
+++ b/src/ap/ieee802_11.h
@@ -239,4 +239,9 @@ void punct_update_legacy_bw(u16 bitmap, u8 pri_chan,
 			    enum oper_chan_width *width, u8 *seg0, u8 *seg1);
 bool hostapd_is_mld_ap(struct hostapd_data *hapd);
 
+#ifdef CONFIG_IEEE80211BE
+struct sta_info *
+hostapd_ml_get_assoc_sta(struct hostapd_data *hapd, struct sta_info *sta,
+			 struct hostapd_data **assoc_hapd);
+#endif
 #endif /* IEEE802_11_H */
diff --git a/src/ap/ieee802_1x.c b/src/ap/ieee802_1x.c
index 052231e34..f977cd5c7 100644
--- a/src/ap/ieee802_1x.c
+++ b/src/ap/ieee802_1x.c
@@ -1142,6 +1142,16 @@ void ieee802_1x_receive(struct hostapd_data *hapd, const u8 *sa, const u8 *buf,
 		return;
 	}
 
+#ifdef CONFIG_IEEE80211BE
+	if (sta->mld_info.mld_sta) {
+		struct hostapd_data *assoc_hapd;
+		struct sta_info *assoc_sta;
+		assoc_sta = hostapd_ml_get_assoc_sta(hapd, sta, &assoc_hapd);
+		sta = assoc_sta;
+		hapd = assoc_hapd;
+	}
+#endif
+
 	if (len < sizeof(*hdr)) {
 		wpa_printf(MSG_INFO, "   too short IEEE 802.1X packet");
 		return;
-- 
2.17.1

