diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 5e5069780..86c50baf1 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1493,7 +1493,7 @@ int hostapd_probe_req_rx(struct hostapd_data *hapd, const u8 *sa, const u8 *da,
 static void hostapd_notify_auth_ft_finish(void *ctx, const u8 *dst,
 					  const u8 *bssid,
 					  u16 auth_transaction, u16 status,
-					  const u8 *ies, size_t ies_len)
+					  const u8 *ies, size_t ies_len, u8 *auth_reply_addr)
 {
 	struct hostapd_data *hapd = ctx;
 	struct sta_info *sta;
@@ -1570,7 +1570,7 @@ static void hostapd_notif_auth(struct hostapd_data *hapd,
 		wpa_ft_process_auth(sta->wpa_sm, rx_auth->bssid,
 				    rx_auth->auth_transaction, rx_auth->ies,
 				    rx_auth->ies_len,
-				    hostapd_notify_auth_ft_finish, hapd);
+				    hostapd_notify_auth_ft_finish, hapd, NULL);
 		return;
 	}
 #endif /* CONFIG_IEEE80211R_AP */
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 5ea8e3d89..7894164ed 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -498,15 +498,21 @@ static int send_auth_reply(struct hostapd_data *hapd, struct sta_info *sta,
 #ifdef CONFIG_IEEE80211R_AP
 static void handle_auth_ft_finish(void *ctx, const u8 *dst, const u8 *bssid,
 				  u16 auth_transaction, u16 status,
-				  const u8 *ies, size_t ies_len)
+				  const u8 *ies, size_t ies_len, u8 *auth_reply_addr)
 {
 	struct hostapd_data *hapd = ctx;
 	struct sta_info *sta;
 	int reply_res;
+	u8 *selected_dst_addr;
 
-	reply_res = send_auth_reply(hapd, NULL, dst, bssid, WLAN_AUTH_FT,
-				    auth_transaction, status, ies, ies_len,
-				    "auth-ft-finish");
+	if (auth_reply_addr)
+		selected_dst_addr = auth_reply_addr;
+	else
+		selected_dst_addr = dst;
+
+	reply_res = send_auth_reply(hapd, NULL, selected_dst_addr, bssid,
+				    WLAN_AUTH_FT, auth_transaction, status,
+				    ies, ies_len, "auth-ft-finish");
 
 	sta = ap_get_sta(hapd, dst);
 	if (sta == NULL)
@@ -726,7 +732,7 @@ static int auth_sae_send_commit(struct hostapd_data *hapd,
 		status = hapd->conf->sae_commit_status;
 	}
 #endif /* CONFIG_TESTING_OPTIONS */
-	reply_res = send_auth_reply(hapd, sta, sta->addr, bssid,
+	reply_res = send_auth_reply(hapd, sta, sta->auth_reply_addr, bssid,
 				    WLAN_AUTH_SAE, 1,
 				    status, wpabuf_head(data),
 				    wpabuf_len(data), "sae-send-commit");
@@ -748,7 +754,7 @@ static int auth_sae_send_confirm(struct hostapd_data *hapd,
 	if (data == NULL)
 		return WLAN_STATUS_UNSPECIFIED_FAILURE;
 
-	reply_res = send_auth_reply(hapd, sta, sta->addr, bssid,
+	reply_res = send_auth_reply(hapd, sta, sta->auth_reply_addr, bssid,
 				    WLAN_AUTH_SAE, 2,
 				    WLAN_STATUS_SUCCESS, wpabuf_head(data),
 				    wpabuf_len(data), "sae-send-confirm");
@@ -2205,7 +2211,8 @@ static void handle_auth_fils_finish(struct hostapd_data *hapd,
 	auth_alg = (pub ||
 		    resp == WLAN_STATUS_FINITE_CYCLIC_GROUP_NOT_SUPPORTED) ?
 		WLAN_AUTH_FILS_SK_PFS : WLAN_AUTH_FILS_SK;
-	send_auth_reply(hapd, sta, sta->addr, hapd->own_addr, auth_alg, 2, resp,
+	send_auth_reply(hapd, sta, sta->auth_reply_addr, hapd->own_addr,
+			auth_alg, 2, resp,
 			data ? wpabuf_head(data) : (u8 *) "",
 			data ? wpabuf_len(data) : 0, "auth-fils-finish");
 	wpabuf_free(data);
@@ -3066,6 +3073,7 @@ static void handle_auth(struct hostapd_data *hapd,
 			goto fail;
 		}
 	}
+    os_memcpy(sta->auth_reply_addr, mgmt->sa, ETH_ALEN);
 
 #ifdef CONFIG_IEEE80211BE
 	if (auth_transaction == 1) {
@@ -3185,7 +3193,7 @@ static void handle_auth(struct hostapd_data *hapd,
 				    auth_transaction, mgmt->u.auth.variable,
 				    len - IEEE80211_HDRLEN -
 				    sizeof(mgmt->u.auth),
-				    handle_auth_ft_finish, hapd);
+				    handle_auth_ft_finish, hapd, mgmt->sa);
 		/* handle_auth_ft_finish() callback will complete auth. */
 		return;
 #endif /* CONFIG_IEEE80211R_AP */
diff --git a/src/ap/sta_info.h b/src/ap/sta_info.h
index e2b9dde87..81ce5205d 100644
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -102,6 +102,7 @@ struct sta_info {
 	struct sta_info *next; /* next entry in sta list */
 	struct sta_info *hnext; /* next entry in hash table list */
 	u8 addr[6];
+    u8 auth_reply_addr[6];
 	be32 ipaddr;
 	struct dl_list ip6addr; /* list head for struct ip6addr */
 	u16 aid; /* STA's unique AID (1 .. 2007) or 0 if not yet assigned */
diff --git a/src/ap/wpa_auth.h b/src/ap/wpa_auth.h
index c076d71a6..deecc4199 100644
--- a/src/ap/wpa_auth.h
+++ b/src/ap/wpa_auth.h
@@ -517,8 +517,8 @@ void wpa_ft_process_auth(struct wpa_state_machine *sm, const u8 *bssid,
 			 u16 auth_transaction, const u8 *ies, size_t ies_len,
 			 void (*cb)(void *ctx, const u8 *dst, const u8 *bssid,
 				    u16 auth_transaction, u16 resp,
-				    const u8 *ies, size_t ies_len),
-			 void *ctx);
+				    const u8 *ies, size_t ies_len, u8 *auth_reply_addr),
+			 void *ctx, u8 *auth_reply_addr);
 int wpa_ft_validate_reassoc(struct wpa_state_machine *sm, const u8 *ies,
 			    size_t ies_len);
 int wpa_ft_action_rx(struct wpa_state_machine *sm, const u8 *data, size_t len);
diff --git a/src/ap/wpa_auth_ft.c b/src/ap/wpa_auth_ft.c
index 4b16f62df..b0fb8ff92 100644
--- a/src/ap/wpa_auth_ft.c
+++ b/src/ap/wpa_auth_ft.c
@@ -3445,8 +3445,8 @@ void wpa_ft_process_auth(struct wpa_state_machine *sm, const u8 *bssid,
 			 u16 auth_transaction, const u8 *ies, size_t ies_len,
 			 void (*cb)(void *ctx, const u8 *dst, const u8 *bssid,
 				    u16 auth_transaction, u16 status,
-				    const u8 *ies, size_t ies_len),
-			 void *ctx)
+				    const u8 *ies, size_t ies_len, u8 *auth_reply_addr),
+			 void *ctx, u8 *auth_reply_addr)
 {
 	u16 status;
 	u8 *resp_ies;
@@ -3480,7 +3480,7 @@ void wpa_ft_process_auth(struct wpa_state_machine *sm, const u8 *bssid,
 		   status2str(status));
 	wpa_hexdump(MSG_DEBUG, "FT: Response IEs", resp_ies, resp_ies_len);
 	cb(ctx, sm->addr, bssid, auth_transaction + 1, status,
-	   resp_ies, resp_ies_len);
+	   resp_ies, resp_ies_len, auth_reply_addr);
 	os_free(resp_ies);
 }
 
@@ -3811,7 +3811,8 @@ int wpa_ft_action_rx(struct wpa_state_machine *sm, const u8 *data, size_t len)
 
 static void wpa_ft_rrb_rx_request_cb(void *ctx, const u8 *dst, const u8 *bssid,
 				     u16 auth_transaction, u16 resp,
-				     const u8 *ies, size_t ies_len)
+				     const u8 *ies, size_t ies_len,
+				     u8 *auth_reply_addr)
 {
 	struct wpa_state_machine *sm = ctx;
 	wpa_printf(MSG_DEBUG, "FT: Over-the-DS RX request cb for " MACSTR,
@@ -4340,7 +4341,7 @@ static void ft_finish_pull(struct wpa_state_machine *sm)
 
 	sm->ft_pending_cb(sm->ft_pending_cb_ctx, sm->addr, sm->wpa_auth->addr,
 			  sm->ft_pending_auth_transaction + 1, status,
-			  resp_ies, resp_ies_len);
+			  resp_ies, resp_ies_len, NULL);
 	os_free(resp_ies);
 }
 
diff --git a/src/ap/wpa_auth_i.h b/src/ap/wpa_auth_i.h
index 74ae5ad2e..6909d6653 100644
--- a/src/ap/wpa_auth_i.h
+++ b/src/ap/wpa_auth_i.h
@@ -140,7 +140,8 @@ struct wpa_state_machine {
 
 	void (*ft_pending_cb)(void *ctx, const u8 *dst, const u8 *bssid,
 			      u16 auth_transaction, u16 status,
-			      const u8 *ies, size_t ies_len);
+			      const u8 *ies, size_t ies_len,
+			      u8 *auth_reply_addr);
 	void *ft_pending_cb_ctx;
 	struct wpabuf *ft_pending_req_ies;
 	u8 ft_pending_pull_nonce[FT_RRB_NONCE_LEN];
-- 
2.17.1

