diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index ff4a8dff6..ac6705bc3 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -723,7 +723,17 @@ sae_confirm_skip_fixed_fields(struct hostapd_data *hapd,
 	 * At this stage we should already have an MLD station and actually SA
 	 * will be replaced with the MLD MAC address by the driver.
 	 */
-	sta = ap_get_sta(hapd, mgmt->sa);
+	//sta = ap_get_sta(hapd, mgmt->sa);
+	for (sta = hapd->sta_list; sta; sta = sta->next) {
+		u8 *link_addr;
+
+		if (!sta->mld_info.mld_sta)
+			continue;
+		link_addr =
+		sta->mld_info.links[sta->mld_assoc_link_id].peer_addr;
+		if (os_memcmp(mgmt->sa, link_addr, ETH_ALEN) == 0)
+			break;
+	}
 	if (!sta) {
 		wpa_printf(MSG_DEBUG, "SAE: No MLD STA for SAE confirm");
 		return NULL;
-- 
2.17.1

