diff --git a/wpa_supplicant/sme.c b/wpa_supplicant/sme.c
index f2281b1..3da5f7e 100644
--- a/wpa_supplicant/sme.c
+++ b/wpa_supplicant/sme.c
@@ -502,7 +502,7 @@ bool wpas_ml_element(struct wpa_supplicant *wpa_s, struct wpa_bss *bss,
 			u8 link_id = *(data + 1) & 0xF;
 
 			if (neigh_bss) {
-				if (wpa_scan_res_match(wpa_s, 0, neigh_bss,
+				if (!wpa_s->current_ssid || wpa_scan_res_match(wpa_s, 0, neigh_bss,
 						       wpa_s->current_ssid,
 						       1, 0)) {
 					wpa_s->valid_links |= BIT(link_id);
