diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 0516213f4..d9e30ece0 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1027,13 +1027,35 @@ void hostapd_event_sta_opmode_changed(struct hostapd_data *hapd, const u8 *addr,
 
 void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 			     int offset, int width, int cf1, int cf2,
-			     u16 punct_bitmap, int finished)
+			     u16 punct_bitmap, int finished, int link_id)
 {
 #ifdef NEED_AP_MLME
 	int channel, chwidth, is_dfs0, is_dfs;
 	u8 seg0_idx = 0, seg1_idx = 0;
 	size_t i;
 
+#ifdef CONFIG_IEEE80211BE
+	if (link_id > 0) {
+		/*
+		 * Get the right interface for the ch-switch event
+		 */
+		struct hostapd_iface *iface = NULL;
+		for (i = 0; i < hapd->iface->interfaces->count; i++) {
+			iface = hapd->iface->interfaces->iface[i];
+
+			if (iface->bss[0]->conf->mld_ap &&
+			    hapd->conf->mld_id == iface->bss[0]->conf->mld_id &&
+			    link_id == iface->bss[0]->mld_link_id) {
+				wpa_printf(MSG_DEBUG, "Switching hapd for %s from "MACSTR" to "MACSTR"\n",
+						__func__, MAC2STR(hapd->own_addr),
+						MAC2STR(iface->bss[0]->own_addr));
+				hapd = iface->bss[0];
+				break;
+			}
+		}
+	}
+#endif
+
 	hostapd_logger(hapd, NULL, HOSTAPD_MODULE_IEEE80211,
 		       HOSTAPD_LEVEL_INFO,
 		       "driver %s channel switch: iface->freq=%d, freq=%d, ht=%d, vht_ch=0x%x, he_ch=0x%x, eht_ch=0x%x, offset=%d, width=%d (%s), cf1=%d, cf2=%d, puncturing_bitmap=0x%x",
@@ -2356,7 +2378,8 @@ void wpa_supplicant_event(void *ctx, enum wpa_event_type event,
 					data->ch_switch.cf1,
 					data->ch_switch.cf2,
 					data->ch_switch.punct_bitmap,
-					event == EVENT_CH_SWITCH);
+					event == EVENT_CH_SWITCH,
+					data->ch_switch.link_id);
 		break;
 	case EVENT_CONNECT_FAILED_REASON:
 		if (!data)
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index 2994b2788..165b0f95d 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -741,7 +741,7 @@ int hostapd_probe_req_rx(struct hostapd_data *hapd, const u8 *sa, const u8 *da,
 			 int ssi_signal);
 void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 			     int offset, int width, int cf1, int cf2,
-			     u16 punct_bitmap, int finished);
+			     u16 punct_bitmap, int finished, int link_id);
 struct survey_results;
 void hostapd_event_get_survey(struct hostapd_iface *iface,
 			      struct survey_results *survey_results);
diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index 9d39703e0..fe5a31d04 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -1269,6 +1269,9 @@ static void mlme_event_ch_switch(struct wpa_driver_nl80211_data *drv,
 	if (finished)
 		bss->flink->freq = data.ch_switch.freq;
 
+	if (!is_sta_interface(drv->nlmode)) {
+		data.ch_switch.link_id = -1;
+	}
 	if (link && is_sta_interface(drv->nlmode)) {
 		u8 link_id = nla_get_u8(link);
 
@@ -1285,6 +1288,10 @@ static void mlme_event_ch_switch(struct wpa_driver_nl80211_data *drv,
 
 		if (link_id != drv->sta_mlo_info.assoc_link_id)
 			return;
+	} else if (link) {
+		u8 link_id = nla_get_u8(link);
+		if (link_id < MAX_NUM_MLD_LINKS)
+			data.ch_switch.link_id = link_id;
 	}
 
 	drv->assoc_freq = data.ch_switch.freq;
diff --git a/wpa_supplicant/ap.c b/wpa_supplicant/ap.c
index ff7c75644..028aca770 100644
--- a/wpa_supplicant/ap.c
+++ b/wpa_supplicant/ap.c
@@ -1853,7 +1853,7 @@ void wpas_ap_ch_switch(struct wpa_supplicant *wpa_s, int freq, int ht,
 		wpa_s->current_ssid->frequency = freq;
 	hostapd_event_ch_switch(iface->bss[0], freq, ht,
 				offset, width, cf1, cf2, punct_bitmap,
-				finished);
+				finished, -1);
 }
 
 
-- 
2.17.1

