diff --git a/wpa_supplicant/sme.c b/wpa_supplicant/sme.c
index de11ea3..dc6bc3f 100644
--- a/wpa_supplicant/sme.c
+++ b/wpa_supplicant/sme.c
@@ -394,6 +394,7 @@ bool wpas_ml_element(struct wpa_supplicant *wpa_s, struct wpa_bss *bss,
 			     BASIC_MULTI_LINK_CTRL_PRES_BSS_PARAM_CH_COUNT |
 			     BASIC_MULTI_LINK_CTRL_PRES_MLD_CAPA);
 	bool ret = false;
+	bool found = false;
 
 	if (!(wpa_s->drv_flags2 & WPA_DRIVER_FLAGS2_MLO) && wpa_s->conf->allow_legacy) {
 		wpa_printf(MSG_ERROR,"%s %d \n",__func__,__LINE__);
@@ -517,11 +518,23 @@ bool wpas_ml_element(struct wpa_supplicant *wpa_s, struct wpa_bss *bss,
 						   "MLD: Neighbor doesn't match current SSID - skip link");
 				}
 			} else {
-				wpa_printf(MSG_DEBUG,
-					   "MLD: Neighbor not found in scan");
-				ret = false;
-				*fail = true;
-				goto out;
+				int partner_freq = ieee80211_chan_to_freq(NULL, ap_info->op_class, ap_info->channel);
+				if (partner_freq && wpa_s->conf->freq_list) {
+					size_t freq_list_size = int_array_len((const int *) wpa_s->conf->freq_list);
+					for ( i = 0; i < freq_list_size; i++) {
+						if (wpa_s->conf->freq_list[i] == partner_freq) {
+							wpa_printf(MSG_DEBUG, "ML Partner freq %d is part of our scan list but still no neighbor info", partner_freq);
+							found = true;
+							break;
+						}
+					}
+				}
+				if (found) {
+					wpa_printf(MSG_DEBUG,"MLD: Neighbor not found in scan");
+					ret = false;
+					*fail = true;
+					goto out;
+				}
 			}
 		}
 
